libs/dmlf/include/dmlf/update.hpp

```diff
@@ -0,0 +1,152 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/serializers/base_types.hpp"
+#include "core/serializers/main_serializer.hpp"
+#include "crypto/hash.hpp"
+#include "crypto/sha256.hpp"
+#include "dmlf/update_interface.hpp"
+
+#include <chrono>
+#include <cstdint>
+#include <vector>
+
+namespace fetch {
+namespace dmlf {
+
+template <typename T>
+class Update : public UpdateInterface
+{
+  template <typename TT, typename D>
+  friend struct serializers::MapSerializer;
+
+public:
+  using TensorType       = T;
+  using VectorTensorType = std::vector<TensorType>;
+  using TimeStampType    = UpdateInterface::TimeStampType;
+  using FingerprintType  = UpdateInterface::FingerprintType;
+
+  using PayloadType = VectorTensorType;
+
+  explicit Update()
+    : stamp_{CurrentTime()}
+  {}
+  explicit Update(VectorTensorType gradients)
+    : stamp_{CurrentTime()}
+    , gradients_{std::move(gradients)}
+    , fingerprint_{ComputeFingerprint()}
+  {}
+
+  byte_array::ByteArray serialise() override
+  {
+    serializers::MsgPackSerializer serializer;
+    serializer << *this;
+    return serializer.data();
+  }
+  byte_array::ByteArray serialise(std::string type) override
+  {
+    serializers::MsgPackSerializer serializer;
+    serializers::MsgPackSerializer serializer_;
+    serializer_ << *this;
+    serializer << type;
+    serializer << serializer_.data();
+    return serializer.data();
+  }
+  void deserialise(const byte_array::ByteArray &map) override
+  {
+    serializers::MsgPackSerializer serializer{map};
+    serializer >> *this;
+  }
+  TimeStampType TimeStamp() const override
+  {
+    return stamp_;
+  }
+  FingerprintType Fingerprint() const override
+  {
+    return fingerprint_;
+  }
+
+  virtual const VectorTensorType &GetGradients() const
+  {
+    return gradients_;
+  }
+
+  ~Update() override = default;
+
+  Update(const Update &other) = delete;
+  Update &operator=(const Update &other)  = delete;
+  bool    operator==(const Update &other) = delete;
+  bool    operator<(const Update &other)  = delete;
+
+protected:
+private:
+  static TimeStampType CurrentTime()
+  {
+    return static_cast<TimeStampType>(std::chrono::duration_cast<std::chrono::milliseconds>(
```

Comment:
You might want to consider looking to the `fetch-moment` library for clocks here. @n-hutton will be doing something similar in the future. The advantage of the `fetch-moment` library is that the clocks are adjustable from tests

Reply:
@ejfitzgerald which clock do you recommend : `GetClock()` or `CreateAdjustableClock()`?

Reply:
Understood. I'll move code to use `GetClock()`